### PR TITLE
Call SetupDevice for Block Volume only if Volume is not globally Mounted

### LINKED ISSUE #112343: CSI `NodeStageVolume` called multiple times for same attached volume on same node
### What happened?

-  For raw block CSI volumes on Kubernetes, kubelet ends up calling CSI `NodeStageVolume` for every single "map" (i.e. raw block "mount") operation for a volume already attached to the node.

### What did you expect to happen?

[Per CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodestagevolume), `NodeStageVolume` should be called once per volume per node:
* More specifically, `This RPC SHOULD be called by the CO when a workload that wants to use the specified volume is placed (scheduled) on the specified node for the first time or for the first time since a NodeUnstageVolume call for the specified volume was called and returned success on that node`.

### How can we reproduce it (as minimally and precisely as possible)?

* Create a raw block PVC for a CSI driver
* Create a pod using that PVC
* Create another pod using the same PVC and schedule to the same node
* Kubelet logs will show multiple `blockMapper.stageVolumeForBlock successfully requested NodeStageVolume` for the same PVC.

Example log:

```
Sep 08 23:27:49 node1 kubelet[2820769]: I0908 23:27:49.078408 2820769 csi_block.go:203] kubernetes.io/csi: blockMapper.stageVolumeForBlock successfully requested NodeStageVolume [/var/lib/kubelet/plugins/
kubernetes.io/csi/volumeDevices/staging/pvc-41f58b43-a3eb-4170-a116-98b301a28a67]
...
Sep 08 23:32:54 node1 kubelet[2820769]: I0908 23:32:54.310885 2820769 csi_block.go:203] kubernetes.io/csi: blockMapper.stageVolumeForBlock successfully requested NodeStageVolume [/var/lib/kubelet/plugins/
kubernetes.io/csi/volumeDevices/staging/pvc-41f58b43-a3eb-4170-a116-98b301a28a67]
```

### Anything else we need to know?

Problem appears to be here:

https://github.com/kubernetes/kubernetes/blob/a7936658baba02e3ee73324ffe670559f03c188a/pkg/volume/util/operationexecutor/operation_generator.go#L1117-L1119

This should first check if "volume is globally mounted" before executing `SetUpDevice`. Does anyone know why we don't? 

For example, if you look a the equivalent code for file-on-block, it does the `DeviceGloballyMounted` check:
https://github.com/kubernetes/kubernetes/blob/a7936658baba02e3ee73324ffe670559f03c188a/pkg/volume/util/operationexecutor/operation_generator.go#L634

CC @mkimuram @vladimirvivien @msau42 @jsafrane @gnufied 

### Kubernetes version

Repro'd on Kubernetes v1.23.5

### Cloud provider

N/A

### OS version

N/A

### Install tools

N/A

### Container runtime (CRI) and version (if applicable)

N/A

### Related plugins (CNI, CSI, ...) and versions (if applicable)

CSI general (not specific to any driver)

### PR DESCRIPTION
**Pull request by akankshakumari393**
#### What type of PR is this?
/kind bug

#### Which issue(s) this PR fixes:

Fixes #112343

#### Does this PR introduce a user-facing change?

```release-note
For raw block CSI volumes on Kubernetes, kubelet was incorrectly calling CSI NodeStageVolume for every single "map" (i.e. raw block "mount") operation for a volume already attached to the node. This PR ensures it is only called once per volume per node.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by saad-ali** [comment]
Do we have sufficient E2E raw block test coverage? I am wondering if there are other bugs that were hidden that will surface once this is fixed.

**Comment by mattcary** [comment]
> Do we have sufficient E2E raw block test coverage? I am wondering if there are other bugs that were hidden that will surface once this is fixed.

Most of the testsuites seem to have a block pattern. So the coverage is probably on par with file---I don't know if that's enough though. There are [tests](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/multivolume.go#L296) for a volume being used multiple times on the same node which [are actually run](https://testgrid.k8s.io/provider-gcp-compute-persistent-disk-csi-driver#Kubernetes%20Master%20Driver%20Latest&include-filter-by-regex=block.*concur), so  I think we might be okay?

```
$ gfind test/e2e storageframework.BlockVol
test/e2e/storage/testsuites/ephemeral.go:71:		storageframework.BlockVolModeGenericEphemeralVolume,
test/e2e/storage/testsuites/volumes.go:88:		storageframework.BlockVolModePreprovisionedPV,
test/e2e/storage/testsuites/volumes.go:89:		storageframework.BlockVolModeDynamicPV,
test/e2e/storage/testsuites/disruptive.go:58:		storageframework.BlockVolModePreprovisionedPV,
test/e2e/storage/testsuites/disruptive.go:59:		storageframework.BlockVolModeDynamicPV,
test/e2e/storage/testsuites/volumemode.go:78:		storageframework.BlockVolModePreprovisionedPV,
test/e2e/storage/testsuites/volumemode.go:79:		storageframework.BlockVolModeDynamicPV,
test/e2e/storage/testsuites/provisioning.go:94:		storageframework.BlockVolModeDynamicPV,
test/e2e/storage/testsuites/volume_expand.go:79:		storageframework.BlockVolModeDynamicPV,
test/e2e/storage/testsuites/volume_expand.go:81:		storageframework.BlockVolModeDynamicPVAllowExpansion,
test/e2e/storage/testsuites/multivolume.go:66:		storageframework.BlockVolModePreprovisionedPV,
test/e2e/storage/testsuites/multivolume.go:67:		storageframework.BlockVolModeDynamicPV,
test/e2e/storage/testsuites/multivolume.go:432:		if l.driver.GetDriverInfo().Name == "vsphere" && pattern == storageframework.BlockVolModeDynamicPV {
test/e2e/storage/testsuites/volume_stress.go:78:		storageframework.BlockVolModeDynamicPV,
```

**Comment by saad-ali** [comment]
Thanks @mattcary. Coverage looks promising. And we are early enough in the release, hopefully if there are any issues, they can be fixed before the 1.26 release.

**Comment by saad-ali** [comment]
/lgtm
/approve

/hold

Adding hold to give others an opportunity to review. e.g. @jsafrane @xing-yang @jingxu97 @msau42

**Comment by saad-ali** [comment]
@akankshakumari393 Would be nice to add a test to prevent regression for this case (ok to do in a follow up PR).

**Comment by akankshakumari393** [comment]
@saad-ali Sure will add a test in a follow up PR.

File: pkg/volume/util/operationexecutor/operation_generator.go [hunk 1]
```diff
@@ -1114,7 +1114,7 @@ func (og *operationGenerator) GenerateMapVolumeFunc(
 
 		}
 		// Call SetUpDevice if blockVolumeMapper implements CustomBlockVolumeMapper
-		if customBlockVolumeMapper, ok := blockVolumeMapper.(volume.CustomBlockVolumeMapper); ok {
+		if customBlockVolumeMapper, ok := blockVolumeMapper.(volume.CustomBlockVolumeMapper); ok && actualStateOfWorld.GetDeviceMountState(volumeToMount.VolumeName) != DeviceGloballyMounted {
```
Comment (msau42) [inline pkg/volume/util/operationexecutor/operation_generator.go:1117]: Can you also test the reconstruction scenario (restart kubelet while the volume is already "mounted")?

Comment (msau42) [reply pkg/volume/util/operationexecutor/operation_generator.go:1117]: Also the scenario where two pods share the same volume on the same node.

Comment (jsafrane) [reply pkg/volume/util/operationexecutor/operation_generator.go:1117]: There is are tests with `[Testpattern: Pre-provisioned PV (block volmode)] multiVolume [Slow] should concurrently access the single volume from pods on the same node` that passed in pull-kubernetes-e2e-gce-storage-slow.

**Comment by jsafrane** [comment]
/test help
checking if we have anything that runs `[Disruptive]` tests.

**Comment by jsafrane** [comment]
Half of block volume tests succeeded and second half failed. It seems that these tests were not run for quite same time.
Trying to get "baseline flakiness" in #111367 and compare with this PR.
/test pull-kubernetes-e2e-gce-storage-disruptive

**Comment by msau42** [comment]
> 

It looks like CSI raw block reconstruction tests passed. The failing raw block tests are related to rbd and iscsi which is potentially an environmental issue

**Comment by msau42** [comment]
/hold cancel
/skip
the failing tests look environmental. The CSI ones passed.

**Comment by jingxu97** [comment]
The change is merged! 
I think we should cherrypick this change to all supporting versions.

**Comment by saad-ali** [comment]
Thanks folks! @akankshakumari393 please link regression test PR when ready here.

**Pull request merged**